// backend/src/trustflow-knowledge/trustflow-knowledge.controller.ts
import { Controller, Post, UploadedFile, UseInterceptors, Param, HttpException, HttpStatus, Get } from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { TrustFlowKnowledgeService } from './trustflow-knowledge.service';
import { PrismaService } from '../prisma/prisma.service';
import { DraftWorker } from '../draft/draft.worker';

@Controller('knowledge-base')
export class TrustFlowKnowledgeController {
  constructor(
    private readonly trustFlowKnowledgeService: TrustFlowKnowledgeService,
    private readonly prisma: PrismaService,
    private readonly draftWorker: DraftWorker,
  ) {}

  @Post(':projectId/ingest')
  @UseInterceptors(FileInterceptor('file'))
  async ingestPdf(@Param('projectId') projectId: string, @UploadedFile() file: Express.Multer.File) {
    if (!file) {
      throw new HttpException('No file uploaded', HttpStatus.BAD_REQUEST);
    }

    if (!file.mimetype.includes('pdf')) {
      throw new HttpException('Only PDF files are supported', HttpStatus.BAD_REQUEST);
    }

    try {
      const numericProjectId = Number(projectId);
      const result = await this.trustFlowKnowledgeService.ingestPdfToKnowledgeBase(
        numericProjectId,
        file.buffer,
        file.originalname,
      );

      await this.prisma.questionItem.updateMany({
        where: {
          projectId: numericProjectId,
          status: {
            in: ['PENDING', 'DRAFTED', 'NEEDS_REVIEW'],
          },
        },
        data: {
          status: 'PENDING',
          answer: null,
          confidence: null,
          citations: null,
        },
      });

      await this.draftWorker.enqueueDraft(numericProjectId);

      return {
        success: true,
        message: `Successfully ingested ${result.chunkCount} chunks and queued draft generation`,
        ...result,
      };
    } catch (error) {
      const message = (error as Error).message || 'Knowledge base ingestion failed';

      if (message.includes('Project not found')) {
        throw new HttpException(message, HttpStatus.NOT_FOUND);
      }

      if (message.includes('SubscriptionRequiredException') || message.includes('AWS Access Key Id needs a subscription')) {
        throw new HttpException(
          'Textract is not enabled for this AWS account/key. Enable Textract access in AWS or use credentials with Textract subscription.',
          HttpStatus.BAD_GATEWAY,
        );
      }

      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Get(':projectId/stats')
  async getKnowledgeStats(@Param('projectId') projectId: string) {
    try {
      return await this.trustFlowKnowledgeService.getKnowledgeStats(Number(projectId));
    } catch (error) {
      const message = (error as Error).message || 'Failed to load knowledge stats';
      if (message.includes('Project not found')) {
        throw new HttpException(message, HttpStatus.NOT_FOUND);
      }
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}
