// backend/src/trustflow-knowledge/trustflow-knowledge.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { AwsIntegrationService } from '../aws-integration/aws-integration.service';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class TrustFlowKnowledgeService {
  private readonly logger = new Logger(TrustFlowKnowledgeService.name);
  private readonly embeddingDimensions = Number(process.env.EMBEDDING_DIM || 1536);
  private readonly chunkSize = Number(process.env.KB_CHUNK_SIZE || 600);
  private readonly chunkOverlap = Number(process.env.KB_CHUNK_OVERLAP || 120);
  private readonly minChunkLength = Number(process.env.KB_MIN_CHUNK_LENGTH || 120);

  constructor(
    private readonly awsIntegrationService: AwsIntegrationService,
    private readonly prisma: PrismaService,
  ) {}

  private toVectorLiteral(embedding: number[]) {
    return `[${embedding.join(',')}]`;
  }

  private normalizeText(rawText: string) {
    return rawText.replace(/\r\n/g, '\n').replace(/\t/g, ' ').trim();
  }

  private getOverlapSuffix(text: string) {
    if (this.chunkOverlap <= 0 || text.length <= this.chunkOverlap) {
      return text;
    }
    return text.slice(text.length - this.chunkOverlap);
  }

  private chunkText(rawText: string) {
    const normalizedText = this.normalizeText(rawText);
    if (!normalizedText) return [];

    const paragraphs = normalizedText
      .split(/\n\s*\n/g)
      .map((paragraph) => paragraph.replace(/\s+/g, ' ').trim())
      .filter(Boolean);

    if (paragraphs.length === 0) return [];

    const chunks: string[] = [];
    let currentChunk = '';

    for (const paragraph of paragraphs) {
      if (paragraph.length >= this.chunkSize) {
        if (currentChunk.trim().length >= this.minChunkLength) {
          chunks.push(currentChunk.trim());
        }
        currentChunk = '';

        for (let start = 0; start < paragraph.length; start += this.chunkSize - this.chunkOverlap) {
          const part = paragraph.slice(start, Math.min(start + this.chunkSize, paragraph.length)).trim();
          if (part.length >= this.minChunkLength) {
            chunks.push(part);
          }
          if (start + this.chunkSize >= paragraph.length) break;
        }
        continue;
      }

      const separator = currentChunk ? ' ' : '';
      const candidate = `${currentChunk}${separator}${paragraph}`;

      if (candidate.length <= this.chunkSize) {
        currentChunk = candidate;
        continue;
      }

      if (currentChunk.trim().length >= this.minChunkLength) {
        chunks.push(currentChunk.trim());
      }

      const overlapSeed = this.getOverlapSuffix(currentChunk.trim());
      currentChunk = overlapSeed ? `${overlapSeed} ${paragraph}` : paragraph;

      if (currentChunk.length > this.chunkSize) {
        const splitHead = currentChunk.slice(0, this.chunkSize).trim();
        if (splitHead.length >= this.minChunkLength) {
          chunks.push(splitHead);
        }
        currentChunk = this.getOverlapSuffix(splitHead);
      }
    }

    if (currentChunk.trim().length >= this.minChunkLength) {
      chunks.push(currentChunk.trim());
    }

    return chunks;
  }

  private async persistEmbedding(projectId: number, chunk: string, source: string) {
    const embedding = await this.awsIntegrationService.generateEmbeddings(chunk, this.embeddingDimensions);
    const vectorExpression = `vector('${this.toVectorLiteral(embedding)}')`;

    await this.prisma.$executeRaw`
      INSERT INTO "Embedding" ("projectId", "chunk", "vector", "source", "createdAt")
      VALUES (${projectId}, ${chunk}, ${Prisma.raw(vectorExpression)}, ${source}, NOW())
    `;
  }

  async ingestPdfToKnowledgeBase(projectId: number, buffer: Buffer, source = 'upload') {
    const project = await this.prisma.project.findUnique({ where: { id: projectId } });
    if (!project) {
      throw new Error('Project not found');
    }

    const parsed = await this.awsIntegrationService.parseDocumentWithTextract(buffer);
    const text = parsed.text;
    if (!text?.trim()) {
      throw new Error('No text found via Textract');
    }

    const chunks = this.chunkText(text);
    if (chunks.length === 0) {
      throw new Error('No chunkable content found in uploaded PDF');
    }

    const results = [];

    for (const chunk of chunks) {
      await this.persistEmbedding(projectId, chunk, source);

      results.push({ chunk, source });
    }

    return {
      projectId,
      source,
      chunkCount: chunks.length,
      inserted: results.length,
      parser: (parsed as { parser?: string }).parser || 'textract',
    };
  }

  async ingestFeedbackAnswer(projectId: number, question: string, answer: string, reviewer = 'human-review') {
    const content = `Question: ${question}\nAnswer: ${answer}`;
    const source = `review-feedback:${reviewer}`;
    await this.persistEmbedding(projectId, content, source);

    this.logger.log(`Feedback embedded for project ${projectId} by ${reviewer}`);
  }

  async getKnowledgeStats(projectId: number) {
    const project = await this.prisma.project.findUnique({ where: { id: projectId } });
    if (!project) {
      throw new Error('Project not found');
    }

    const chunkCount = await this.prisma.embedding.count({
      where: { projectId },
    });

    const latestEmbeddings = await this.prisma.embedding.findMany({
      where: { projectId },
      orderBy: { createdAt: 'desc' },
      take: 10,
      select: { source: true },
    });

    const sources = Array.from(
      new Set(
        latestEmbeddings
          .map((item) => item.source)
          .filter((value): value is string => Boolean(value)),
      ),
    );

    return {
      projectId,
      chunkCount,
      sourceCount: sources.length,
      recentSources: sources,
    };
  }
}
