// backend/src/draft/draft.worker.ts
import { Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { Worker, Queue } from 'bullmq';
import { AwsIntegrationService } from '../aws-integration/aws-integration.service';
import { PrismaService } from '../prisma/prisma.service';

interface CitationItem {
  embeddingId: number;
  score: number;
  snippet: string;
  source: string;
}

@Injectable()
export class DraftWorker implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(DraftWorker.name);
  private readonly queueName = 'draft-questions';
  private readonly embeddingDimensions = Number(process.env.EMBEDDING_DIM || 1536);
  private readonly retrievalTopK = Number(process.env.RETRIEVAL_TOP_K || 6);
  private queue: Queue;
  private worker: Worker;

  constructor(
    private readonly prisma: PrismaService,
    private readonly awsIntegrationService: AwsIntegrationService,
  ) {
    const connection = {
      host: process.env.REDIS_HOST || 'localhost',
      port: Number(process.env.REDIS_PORT || 6379),
    };

    this.queue = new Queue(this.queueName, { connection });

    this.worker = new Worker(
      this.queueName,
      async (job) => {
        const { projectId } = job.data;
        return this.processDraftQuestions(projectId);
      },
      { connection },
    );

    this.worker.on('completed', (job) => {
      this.logger.log(`Draft job completed for project ${job.data.projectId}`);
    });

    this.worker.on('failed', (job, err) => {
      this.logger.error(`Draft job failed for project ${job?.data?.projectId}`, err);
    });
  }

  async onModuleInit() {
    await this.queue.waitUntilReady();
    await this.worker.waitUntilReady();

    this.logger.log('DraftWorker initialized and ready');
  }

  async onModuleDestroy() {
    await this.worker.close();
    await this.queue.close();
  }

  async enqueueDraft(projectId: number, opts?: any) {
    return this.queue.add('process', { projectId }, opts ?? {});
  }

  private async fetchTopChunks(projectId: number, embedding: number[]) {
    const vectorLiteral = `[${embedding.join(',')}]`;
    const safeProjectId = Number(projectId);
    const query = `
      SELECT "id", "chunk", "source", (1.0 / (1.0 + ("vector" <=> vector('${vectorLiteral}')))) AS similarity
      FROM "Embedding"
      WHERE "projectId" = ${safeProjectId}
      ORDER BY "vector" <=> vector('${vectorLiteral}')
      LIMIT ${Math.max(1, this.retrievalTopK)};
    `;
    const rows: Array<{ id: number; chunk: string; source: string | null; similarity: number }> =
      await this.prisma.$queryRawUnsafe(query);

    return rows;
  }

  public async processDraftQuestions(projectId: number) {
    const pendingItems = await this.prisma.questionItem.findMany({
      where: { projectId, status: 'PENDING' },
    });

    if (pendingItems.length === 0) {
      this.logger.log(`No pending questions for project ${projectId}`);
      return { processed: 0 };
    }

    for (const question of pendingItems) {
      try {
        const contextEmbeddings = await this.awsIntegrationService.generateEmbeddings(
          question.question,
          this.embeddingDimensions,
        );
        const topChunks = await this.fetchTopChunks(projectId, contextEmbeddings);

        if (topChunks.length === 0) {
          await this.prisma.questionItem.update({
            where: { id: question.id },
            data: {
              answer: 'Not enough information',
              confidence: 0,
              status: 'NEEDS_REVIEW',
              citations: JSON.stringify([]),
            },
          });
          continue;
        }

        const citations: CitationItem[] = topChunks.map((chunk) => ({
          embeddingId: chunk.id,
          score: Number(chunk.similarity),
          snippet: chunk.chunk,
          source: chunk.source ?? 'unknown',
        }));

        const contextText = citations
          .map(
            (c, i) =>
              `Citation ${i + 1} | source=${c.source} | score=${c.score.toFixed(4)}\n${c.snippet}`,
          )
          .join('\n\n');

        const prompt = `You are a GRC compliance analyst.

Question:
${question.question}

Context:
${contextText}

Rules:
- Use only the provided context.
- If context is insufficient, say "Not enough information".
- Provide a concise, audit-friendly answer.
- End with a short "Citations:" line listing the citation numbers you used.`;

        const aiAnswer = await this.awsIntegrationService.generateDraftAnswer(prompt);
        const normalizedAnswer = aiAnswer.trim().toLowerCase();
        const contextHasSubstance = contextText.length >= 500;
        const fallbackAnswer = `${citations[0].snippet.slice(0, 650).trim()}\n\nCitations: 1`;
        const finalAnswer =
          normalizedAnswer === 'not enough information' && contextHasSubstance ? fallbackAnswer : aiAnswer;

        const highestSimilarity = topChunks.length > 0 ? Math.max(...topChunks.map((c) => c.similarity)) : 0;
        const status = highestSimilarity < 0.65 ? 'NEEDS_REVIEW' : 'DRAFTED';

        await this.prisma.questionItem.update({
          where: { id: question.id },
          data: {
            answer: finalAnswer,
            confidence: highestSimilarity,
            status,
            citations: JSON.stringify(citations),
          },
        });
      } catch (err) {
        this.logger.error(`Error processing question ${question.id}`, err as Error);
      }
    }

    return { processed: pendingItems.length };
  }
}
